mmcv/ops/csrc/onnxruntime/cpu/modulated_deform_conv.cpp

```diff
@@ -0,0 +1,311 @@
+
+    float val = (w1 * v1 + w2 * v2 + w3 * v3 + w4 * v4);
+    return val;
+}
+
+// output: (channels * kernel_h * kernel_w, dst_h * dst_w)
+void deformable_im2col_2d(
+    const float* input,
+    const float* offset,
+    const float* mask,
+    const int64_t src_h,
+    const int64_t src_w,
+    const int64_t kernel_h,
+    const int64_t kernel_w,
+    const int64_t pad_h,
+    const int64_t pad_w,
+    const int64_t stride_h,
+    const int64_t stride_w,
+    const int64_t dilation_h,
+    const int64_t dilation_w,
+    const int64_t channels,
+    const int64_t offset_groups,
+    const int64_t dst_h,
+    const int64_t dst_w,
+    const bool use_mask,
+    float* columns)
+{
+    const int64_t workload = channels * dst_h * dst_w;
+    for (int64_t index = 0; index != workload; ++index) {
+        const int64_t ow = index % dst_w;
+        const int64_t oh = (index / dst_w) % dst_h;
+        const int64_t ic = index / (dst_w * dst_h);
+        const int64_t oc = ic * kernel_h * kernel_w;
+
+        int64_t c_per_offset_grp = channels / offset_groups;
+        const int64_t grp_idx = ic / c_per_offset_grp;
+
+        auto columns_ptr = columns + (oc * (dst_h * dst_w) + oh * dst_w + ow);
+        auto input_ptr = input + ic * (src_h * src_w);
+        auto offset_ptr = offset + grp_idx * 2 * kernel_h * kernel_w * dst_h * dst_w;
+        auto mask_ptr = mask;
+        if (use_mask) {
+            mask_ptr += grp_idx * kernel_h * kernel_w * dst_h * dst_w;
+        }
+
+        for (int64_t kh = 0; kh < kernel_h; ++kh) {
+            for (int64_t kw = 0; kw < kernel_w; ++kw) {
+                const int64_t mask_idx = kh * kernel_w + kw;
+                const int64_t offset_idx = 2 * mask_idx;
+
+                float mask_value = 1;
+                if (use_mask) {
+                mask_value =
+                    mask_ptr[mask_idx * (dst_h * dst_w) + oh * dst_w + ow];
+                }
+
+                const float offset_h = offset_ptr[offset_idx * (dst_h * dst_w) + oh * dst_w + ow];
+                const float offset_w = offset_ptr[(offset_idx + 1) * (dst_h * dst_w) + oh * dst_w + ow];
+                const float ih = (oh * stride_h - pad_h) + kh * dilation_h + offset_h;
+                const float iw = (ow * stride_w - pad_w) + kw * dilation_w + offset_w;
+                *columns_ptr = mask_value * bilinear_interpolate_2d(input_ptr, src_h, src_w, ih, iw);
+                columns_ptr += dst_h * dst_w;
+            }
+        }
+    }
+}
+
+void gemm_ref_fp32(
+    const float *A,
+    const float *B,
+    const float *V,
+    const float *H,
+    const int32_t trans_A,
+    const int32_t trans_B,
+    const int32_t M,
+    const int32_t N,
+    const int32_t K,
+    const float alpha,
+    const float beta,
+    float *Y)
+{
+    for (int64_t m = 0; m < M; ++m) {
+        for (int64_t n = 0; n < N; ++n) {
+            float y = 0.0f;
+            if (!trans_A && !trans_B) { // MK, KN; NN
```

Comment:
Can we move the condition line outside the loop funciton of gemm? AFAK, a inside `if else` may slow a loop a lot compared with outside.